GitTfs/Core/GitRepository.cs

```diff
@@ -508,9 +513,30 @@ public string AssertValidBranchName(string gitBranchName)
         {
             if (!Reference.IsValidName(ShortToLocalName(gitBranchName)))
                 throw new GitTfsException("The name specified for the new git branch is not allowed. Choose another one!");
+            var refs = _repository.Refs.Select(r => r.CanonicalName).ToList();
+            while (IsRefNameUsed(gitBranchName, refs))
+            {
+                gitBranchName = "_" + gitBranchName;
+            }
             return gitBranchName;
         }
 
+        private static bool IsRefNameUsed(string gitBranchName, List<string> refs)
+        {
+            var parts = gitBranchName.Split('/');
+            var refName = parts.First();
+            for (int i = 1; i <= parts.Length; i++)
+            {
+                if (refs.Exists(s=>string.Compare(ShortToLocalName(refName), s, StringComparison.InvariantCultureIgnoreCase) == 0)
+                    || refs.Exists(s => string.Compare(ShortToTfsRemoteName(refName), s, StringComparison.InvariantCultureIgnoreCase) == 0))
```

Comment:
Why not just ask `_repository.Refs`? I haven't looked, but I assume it has an API to check if a ref exists. It seems like it would avoid a bunch of case comparisons if this was something like

```csharp
if (_repository.refs.Exist(ShortToLocalName(refName)) || _repository.refs.Exist(ShortToTfsRemoteName(refName)))
  return true;
```